### PR TITLE
Issue #2630366 by andrewbelcher: Optionally Unique E-mails

### PR DESCRIPTION
**Pull request by andrewbelcher**

### CODE REVIEW

File: tests/src/Kernel/UniqueEmailTest.php [hunk 1]
```diff
@@ -0,0 +1,586 @@
...
+   * @covers ::validate
+   */
+  public function testModeAllCoupled() {
+    $this->setUniqueEmailsConfig('all');
+
+    $existing_user = $this->createUser();
+
+    // Test validating a decoupled user.
+    // Expected result: Violations
+    $user = $this->createUnsavedUser(TRUE, $existing_user->getAccountName());
```
Comment (rlmumford) [inline tests/src/Kernel/UniqueEmailTest.php]: Why does this line use getAccountName() but line 96 above use email_prefix?

File: tests/src/Kernel/UniqueEmailTest.php [hunk 2]
```diff
@@ -0,0 +1,586 @@
...
+  public function testModeNoneCoupled() {
+    $this->setUniqueEmailsConfig('none');
+
+    $existing_user = $this->createUser();
+
+    // Test validating a decoupled user.
+    // Expected result: Violations
+    $user = $this->createUnsavedUser(TRUE, $existing_user->getAccountName());
+    $violations = $user->validate();
+    $this->assertNotEmpty($violations, 'Decoupled user fails validation.');
```
Comment (rlmumford) [inline tests/src/Kernel/UniqueEmailTest.php]: Shouldn't this be asserting empty? If the mode is none, then a decoupled user can have the same email address as a coupled user.

File: tests/src/Kernel/UniqueEmailTest.php [hunk 3]
```diff
@@ -0,0 +1,586 @@
...
+    $user = $this->createUnsavedUser(TRUE, $existing_user->email_prefix);
+    $violations = $user->validate();
+    $this->assertEmpty($violations, 'Decoupled user without role passes validation.');
+
+    // Test validating a decoupled user with the role.
+    // Expected result: No violations
+    $user = $this->createUnsavedUser(TRUE, $existing_user->email_prefix);
+    $user->addRole($role);
+    $violations = $user->validate();
+    $this->assertEmpty($violations, 'Decoupled user with role passes validation.');
```
Comment (rlmumford) [inline tests/src/Kernel/UniqueEmailTest.php:285]: --If the user has the role, it's email SHOULD be unique, so this should be assertNotEmpty()--

Ignore this. The existing user does not have the role.

File: tests/src/Kernel/UniqueEmailTest.php [hunk 4]
```diff
@@ -0,0 +1,586 @@
...
+    $user = $this->createUnsavedUser(TRUE, $existing_user->email_prefix);
+    $user->addRole($role);
+    $violations = $user->validate();
+    $this->assertNotEmpty($violations, 'Decoupled user with role fails validation.');
+
+    // Test validating a coupled user without the role.
+    // Expected result: Violations
+    $user = $this->createUnsavedUser(FALSE, $existing_user->email_prefix);
+    $violations = $user->validate();
+    $this->assertNotEmpty($violations, 'Coupled user without the role fails validation.');
```
Comment (rlmumford) [inline tests/src/Kernel/UniqueEmailTest.php:333]: As the existing user has the role, and the new user doesn't, shouldn't this be assertEmpty?

File: tests/src/Kernel/UniqueEmailTest.php [hunk 5]
```diff
@@ -0,0 +1,586 @@
...
+   *
+   * @covers ::validate
+   */
+  public function testModeIncludeCoupledWithoutRole() {
+    $role = $this->createRole([]);
+    $this->setUniqueEmailsConfig('include', [$role]);
+
+    $existing_user = $this->createUser();
+
+    // Test validating a coupled user without the role.
```
Comment (rlmumford) [inline tests/src/Kernel/UniqueEmailTest.php]: Comment is wrong. The following line creates a decoupled user.

File: tests/src/Kernel/UniqueEmailTest.php [hunk 6]
```diff
@@ -0,0 +1,586 @@
...
+    // Expected result: No violations
+    $user = $this->createUnsavedUser(TRUE, $existing_user->getAccountName());
+    $violations = $user->validate();
+    $this->assertEmpty($violations, 'Decoupled user without role passes validation.');
+
+    // Test validating a coupled user with the role.
```
Comment (rlmumford) [inline tests/src/Kernel/UniqueEmailTest.php]: Comment is wrong. The following line creates a decoupled user.

File: tests/src/Kernel/UniqueEmailTest.php [hunk 7]
```diff
@@ -0,0 +1,586 @@
...
+    $role = $this->createRole([]);
+    $this->setUniqueEmailsConfig('include', [$role]);
+
+    $existing_user = $this->createUser();
+    $existing_user->addRole($role);
+    $existing_user->save();
+
+    // Test validating a coupled user without the role.
+    // Expected result: No violations
+    $user = $this->createUnsavedUser(TRUE, $existing_user->getAccountName());
```
Comment (rlmumford) [inline tests/src/Kernel/UniqueEmailTest.php]: Comment and message is wrong? OR the createUnsavedUser call is wrong? It says its testing a coupled user but is actually testing a decoupled.

**Comment by rlmumford** [comment]
I reviewed all the tests up to and including the 'include' mode tests, and you can see my comments above. I haven't looked at the exclude mode tests yet.

I am finding the code generally difficult to read and I have a couple of suggestions to make it easier:
1. Rather than supplying TRUE/FALSE to createUnsavedUser we could have the Trait provide the constants USER_DECOUPLED and USER_COUPLED? This would instantly make it easier to read and see what sort of user you were creating. (True/False is _especially_ confusing when TRUE => 'Not Coupled' and FALSE => 'Coupled')
2. Consider renaming the modes/giving these constants too? I find myself trying to think hard about what include/exclude mean on top of remembering that 'none' actually means 'only coupled'. Perhaps we could rename them as follows:
   - 'none' => 'only_coupled'
   - 'include' => 'only_with_role'
   - 'exclude' => 'only_without_role'

**Comment by andrewbelcher** [comment]
Constants are a great idea. I'll get those in. I'll have a closer look through the assert questions later.

File: tests/src/Kernel/UniqueEmailTest.php (hunk 1 above)
Comment (andrewbelcher) [inline tests/src/Kernel/UniqueEmailTest.php]: Decoupled users don't have `name` and the email is suffixed with `@example.com`, so `createDecoupledUser` stores the email prefix against the entity so we can retrieve it later without having to strip off the `@example.com`.

File: tests/src/Kernel/UniqueEmailTest.php (hunk 2 above)
Comment (andrewbelcher) [inline tests/src/Kernel/UniqueEmailTest.php]: Yes, you're right!

File: tests/src/Kernel/UniqueEmailTest.php (hunk 4 above)
Comment (andrewbelcher) [inline tests/src/Kernel/UniqueEmailTest.php:333]: Correct again!

Comment (andrewbelcher) [inline tests/src/Kernel/UniqueEmailTest.php:333]: Actually, not right as this test is in 'include' mode, so the existing user reserves it and as we are creating a coupled user it needs to be unique.

**Pull request merged**